build/xtask/src/sizes.rs

```diff
@@ -79,233 +68,235 @@ pub fn run(
-            o => bail!("Invalid Object {:?}", o),
-        };
-
-        let mut memory_sizes = IndexMap::new();
-        for phdr in &elf.program_headers {
-            if let Some(vregion) = output_region(phdr.p_vaddr) {
-                *memory_sizes.entry(vregion).or_default() += phdr.p_memsz;
-            }
-            // If the VirtAddr disagrees with the PhysAddr, then this is a
-            // section which is relocated into RAM, so we also accumulate
-            // its FileSiz in the physical address (which is presumably
-            // flash).
-            if phdr.p_vaddr != phdr.p_paddr {
-                let region = output_region(phdr.p_paddr).unwrap();
-                *memory_sizes.entry(region).or_default() += phdr.p_filesz;
-            }
+    // We can't naively add up section sizes, since there may be gaps left
+    // by alignment requirements.  Instead, we track the min and max bounds
+    // within each memory region (flash, RAM, etc), then extract the sizes
+    // afterwards.
+    let mut memory_sizes = IndexMap::new();
+    let mut record_size = |start, size| {
+        if let Some(region) = toml.output_region(start) {
+            let end = start + size;
+            let r = memory_sizes.entry(region).or_insert_with(|| start..end);
+            r.start = r.start.min(start);
+            r.end = r.end.max(end);
+            true
+        } else {
+            false
         }
-        *memory_sizes.entry("ram").or_default() += stacksize as u64;
-
-        let mut my_suggestions = Vec::new();
-
-        for (mem_name, used) in memory_sizes {
-            if let Some(&size) = requires.get(mem_name) {
-                sizes.insert(mem_name.to_string(), used);
-                let suggestion = suggest(used);
-                if suggestion < size as u64 {
-                    my_suggestions.push((
-                        mem_name.to_string(),
-                        size,
-                        suggestion,
-                    ));
-                }
-            } else {
-                assert!(used == 0);
+    };
+    for phdr in &elf.program_headers {
+        record_size(phdr.p_vaddr, phdr.p_memsz);
+
+        // If the VirtAddr disagrees with the PhysAddr, then this is a
+        // section which is relocated into RAM, so we also accumulate
+        // its FileSiz in the physical address (which is presumably
+        // flash).
+        if phdr.p_vaddr != phdr.p_paddr {
+            if !record_size(phdr.p_paddr, phdr.p_filesz) {
+                bail!("Failed to remap relocated section at {}", phdr.p_paddr);
             }
         }
-        if !my_suggestions.is_empty() {
-            suggestions.push(my_suggestions);
-        }
-        Ok((sizes, suggestions))
-    };
+    }
+    let mut memory_sizes: IndexMap<&str, u64> = memory_sizes
+        .into_iter()
+        .map(|(name, range)| (name, range.end - range.start))
+        .collect();
+
+    // XXX: are there alignment issues with the stack here?
```

Comment:
Yes, the stack is 8-byte aligned on all of our current targets -- so, technically, this could underestimate RAM usage by 7 bytes in corner cases.